### PR TITLE
Add Checklists Feature

### PR DESCRIPTION
**Pull request by roundhill**
Adds a UI for "checkable" checklists in the note editor. Behind the scenes, we parse markdown checklist syntax (e.g. "- [ ]") and convert them to ClickableSpans. When the note saves, it parses the spannable string and converts it back to the markdown syntax. The app will also display the checklist icons in the notes list.

screen shot 2018-12-21 at 12 08 04 pm

Tapping on a checklist item will add or remove the `x` to the note content when it saves.

There were some unfortunate rabbit holes when making this :) There's some new subclasses:

* `CheckableSpan`: Includes a boolean to track whether the checkbox is ticked or not and to display the appropriate drawable.
* `SimplenoteMovementMethod`: Allows the `CheckableSpan`s to be tapped, but doesn't fire the onClick() listener for any other spans. This is so the regular link detector still works as expected (placing the cursor inside a web url for example should show the context action buttons in the toolbar)
* `CenteredImageSpan`: Centers the drawable with the text, so the checkbox aligns nicely.
* There's also a lot of work to figure out how to highlight the search results properly. We need to adjust the index of the matches so that they still highlight correctly after the note content has been replaced with the `CheckableSpan`s.

**To Test**
* Open a note, and create a checklist! Pressing `return` at the end of a checklist item should automatically add a new one on the next line.
* Open the note on app.simplenote.com, it should have the correct markdown syntax displayed.
* Update the note on the web browser, the app should update and display any modified checklists appropriately.
* Test things that modify the note content, like the history browser and ensure they work as expected.
* Run the new tests!

Refs: #528 #605

### CODE REVIEW

**Comment by roundhill** [comment]
> When I type the closing "]" after entering "- [ " before it, the app stops responding and crashes with the stack trace below.

It's an infinite loop 👻 `setText` fires the `TextWatcher` endlessly. I think I can get away without using `setText` here to fix it.

**Comment by roundhill** [comment]
@theck13 ok, ready for another look.

**Comment by CryptoStar2000** [comment]
Is there just any way to contribute to the test, too? I would love to test the beta :-). . Could you send it to my username plus "@quantentunnel.de" ? thx!!

**Comment by roundhill** [comment]
> Is there just any way to contribute to the test, too? I would love to test the beta :-). . Could you send it to my username plus "@quantentunnel.de" ? thx!!

Yeah I'll post an apk to a release here on GitHub soon for you to try out.

**Comment by CryptoStar2000** [comment]
Thx Dan :-)!! The feature is awesome!! 

I have just a few little comments:

It woukd be perfect if the checkbox-behaviour would be analogue to the 'bullets' like the '-'. Meaning that since intended bullets are recognized, intended checkboxes should also be. This would enable sub-checklists too, as possible with bullets.
If you start a sub-checkbox with some space in front of the box and access the note again, its not recognized as a touch checkbox but as e.g. '   - [ ] Test'

Finally: If you push the checkbox button in an empty line, the cursor is left to the checkbox. For being able to directly start writing, it would be better to have the cursor on the right side of the box.

**Comment by roundhill** [comment]
> If you push the checkbox button in an empty line, the cursor is left to the checkbox. For being able to directly start writing, it would be better to have the cursor on the right side of the box.

Fixed that, thank you!

**Comment by roundhill** [comment]
> It woukd be perfect if the checkbox-behaviour would be analogue to the 'bullets' like the '-'. Meaning that since intended bullets are recognized, intended checkboxes should also be. This would enable sub-checklists too, as possible with bullets.
> If you start a sub-checkbox with some space in front of the box and access the note again, its not recognized as a touch checkbox but as simple text with some space characters in front to the text:
> 
> '(x space characters) - [ ] Test' => no touch checkbox

I think we'll have to leave this as-is for now, it only checks for checklists that start on a newline with `- [  ]` or `- [x]`, to match the markdown syntax

**Comment by CryptoStar2000** [comment]
Wonderful  - thx :-)!

**Comment by roundhill** [comment]
Thanks @theck13! I'll get right on those fixes.

**Comment by roundhill** [comment]
> The ic_checked.xml drawable appears to be taken from gridicons-checkmark-circle. That icon seemed odd to me for a few reasons. 

@drw158 do you have any thoughts on the different checkbox icons for Android? I agree with Tyler that the square version looks better, at least on Android.

https://user-images.githubusercontent.com/3827611/51157776-72433180-184f-11e9-8860-40db102adba5.png

**Comment by roundhill** [comment]
> The search results show notes with checklists containing a checked item when using "x" as the search query even if there is no "x" in the text. See the screenshot below for illustration.

Unfortunately I can't think of an easy fix for this... will 🤔

**Comment by CryptoStar2000** [comment]
I would prefer the square :-)

File: Simplenote/src/main/java/com/automattic/simplenote/utils/ThemeUtils.java [hunk 1]
```diff
@@ -83,4 +87,19 @@ public static int getOptimalDrawerWidth(Context context) {
         int maxPx = DisplayUtils.dpToPx(context, maxDp);
         return Math.min(drawerWidth, maxPx);
     }
+
+    public static int getThemeTextColorId(Context context) {
+        if (context == null) {
+            return 0;
+        }
+
+        if (mTextColorId == 0) {
```
[the remark this answers is not in the document]
Comment (roundhill) [reply Simplenote/src/main/java/com/automattic/simplenote/utils/ThemeUtils.java]: Nice catch! Performance seems fine, so I removed it.

**Comment by roundhill** [comment]
> It seems like something that users could encounter if they don't select multiple lines to insert checkboxes to an entire group at once. What do you think?

Oh, I missed that one. I agree that it should be fixed. I'll take a look at fixing it now.

**Comment by roundhill** [comment]
> Oh, I missed that one. I agree that it should be fixed. I'll take a look at fixing it now.

Fixed in ae83c81. Give it (hopefully) once more run through!

**Comment by theck13** [comment]
> It seems like something that users could encounter if they don't select multiple lines to insert checkboxes to an entire group at once. What do you think?

That's fixed!  I forgot to mention one more thing that might have been missed.  Here it is below.

<blockquote>
If a note contains a checklist long enough to wrap to a second line, the content preview in the note list hides the first checkbox drawable on the second line. See the screenshot below for illustration.<br><br>

simplenote_note_list_wrapping
</blockquote>

Notice the "Sixth" checklist item does not have the checkbox drawable in the list preview.  The screenshot is from testing earlier in this pull request, but the issue still remains.  Do you think we should fix it with these changes or in a separate pull request?  It's a bug, but it's tiny and not a showstopper in my opinion.

**Comment by roundhill** [comment]
> Notice the "Sixth" checklist item does not have the checkbox drawable in the list preview.

Oh yes, I forgot to mention that I looked into this and cannot figure out what's going on there. It could be deep down in the framework and how it handles line wrapping with spans? How about I create an issue for it :)

**Comment by theck13** [comment]
> How about I create an issue for it :)
Sound good to me.

**Pull request merged**